File: SRPM-master/API/Services/ProjectService.cs
using SRPM.API.Models;
using SRPM.Data.Entities;
using SRPM.Data.Repositories;

namespace SRPM.API.Services
{
    public class ProjectService : IProjectService
    {
        private readonly IProjectRepository _projectRepository;
        private readonly IUserRepository _userRepository;
        private readonly INotificationRepository _notificationRepository;

        public ProjectService(
            IProjectRepository projectRepository,
            IUserRepository userRepository,
            INotificationRepository notificationRepository)
        {
            _projectRepository = projectRepository;
            _userRepository = userRepository;
            _notificationRepository = notificationRepository;
        }

        public async Task<ProjectDto?> GetByIdAsync(int id, int userId)
        {
            var project = await _projectRepository.GetByIdAsync(id);
            if (project == null)
                return null;

            // Check if user is a member of the project or the owner
            bool isMember = project.OwnerId == userId || project.ProjectMembers.Any(pm => pm.UserId == userId);
            
            var projectDto = new ProjectDto
            {
                Id = project.Id,
                Title = project.Title,
                Description = project.Description,
                Status = project.Status,
                StartDate = project.StartDate,
                EndDate = project.EndDate,
                OwnerId = project.OwnerId,
                OwnerName = project.Owner.Name,
                ResearchTopicId = project.ResearchTopicId,
                ResearchTopicTitle = project.ResearchTopic?.Title,
                CreatedAt = project.CreatedAt,
                UpdatedAt = project.UpdatedAt
            };

            // If user is not a member, return limited information
            if (!isMember)
            {
                return new ProjectDto
                {
                    Id = project.Id,
                    Title = project.Title,
                    Description = project.Description,
                    Status = project.Status,
                    OwnerId = project.OwnerId,
                    OwnerName = project.Owner.Name
                };
            }

            // Return full information for members
            return projectDto;
        }

        public async Task<IEnumerable<ProjectDto>> GetAllAsync(int userId)
        {
            var projects = await _projectRepository.GetAllAsync();
            var projectDtos = new List<ProjectDto>();

            foreach (var project in projects)
            {
                // Check if user is a member of the project or the owner
                bool isMember = project.OwnerId == userId || project.ProjectMembers.Any(pm => pm.UserId == userId);

                var projectDto = new ProjectDto
                {
                    Id = project.Id,
                    Title = project.Title,
                    Description = project.Description,
                    Status = project.Status,
                    OwnerId = project.OwnerId,
                    OwnerName = project.Owner.Name
                };

                // Add more details if user is a member
                if (isMember)
                {
                    projectDto.StartDate = project.StartDate;
                    projectDto.EndDate = project.EndDate;
                    projectDto.ResearchTopicId = project.ResearchTopicId;
                    projectDto.ResearchTopicTitle = project.ResearchTopic?.Title;
                    projectDto.CreatedAt = project.CreatedAt;
                    projectDto.UpdatedAt = project.UpdatedAt;
                }

                projectDtos.Add(projectDto);
            }

            return projectDtos;
        }

        public async Task<IEnumerable<ProjectDto>> GetByOwnerIdAsync(int ownerId)
        {
            var projects = await _projectRepository.GetByOwnerIdAsync(ownerId);
            return projects.Select(p => new ProjectDto
            {
                Id = p.Id,
                Title = p.Title,
                Description = p.Description,
                Status = p.Status,
                StartDate = p.StartDate,
                EndDate = p.EndDate,
                OwnerId = p.OwnerId,
                OwnerName = p.Owner.Name,
                ResearchTopicId = p.ResearchTopicId,
                ResearchTopicTitle = p.ResearchTopic?.Title,
                CreatedAt = p.CreatedAt,
                UpdatedAt = p.UpdatedAt
            });
        }

        public async Task<IEnumerable<ProjectDto>> GetByMemberIdAsync(int memberId)
        {
            var projects = await _projectRepository.GetByMemberIdAsync(memberId);
            return projects.Select(p => new ProjectDto
            {
                Id = p.Id,
                Title = p.Title,
                Description = p.Description,
                Status = p.Status,
                StartDate = p.StartDate,
                EndDate = p.EndDate,
                OwnerId = p.OwnerId,
                OwnerName = p.Owner.Name,
                ResearchTopicId = p.ResearchTopicId,
                ResearchTopicTitle = p.ResearchTopic?.Title,
                CreatedAt = p.CreatedAt,
                UpdatedAt = p.UpdatedAt
            });
        }

        public async Task<IEnumerable<ProjectDto>> GetByStatusAsync(string status)
        {
            var projects = await _projectRepository.GetByStatusAsync(status);
            return projects.Select(p => new ProjectDto
            {
                Id = p.Id,
                Title = p.Title,
                Description = p.Description,
                Status = p.Status,
                StartDate = p.StartDate,
                EndDate = p.EndDate,
                OwnerId = p.OwnerId,
                OwnerName = p.Owner.Name,
                ResearchTopicId = p.ResearchTopicId,
                ResearchTopicTitle = p.ResearchTopic?.Title,
                CreatedAt = p.CreatedAt,
                UpdatedAt = p.UpdatedAt
            });
        }

        public async Task<IEnumerable<ProjectDto>> GetByResearchTopicIdAsync(int researchTopicId)
        {
            var projects = await _projectRepository.GetByResearchTopicIdAsync(researchTopicId);
            return projects.Select(p => new ProjectDto
            {
                Id = p.Id,
                Title = p.Title,
                Description = p.Description,
                Status = p.Status,
                StartDate = p.StartDate,
                EndDate = p.EndDate,
                OwnerId = p.OwnerId,
                OwnerName = p.Owner.Name,
                ResearchTopicId = p.ResearchTopicId,
                ResearchTopicTitle = p.ResearchTopic?.Title,
                CreatedAt = p.CreatedAt,
                UpdatedAt = p.UpdatedAt
            });
        }

        public async Task<ProjectDto?> CreateAsync(CreateProjectRequest request, int ownerId)
        {
            var owner = await _userRepository.GetByIdAsync(ownerId);
            if (owner == null)
                return null;

            // Check if user has Principal Investigator role
            var roles = await _userRepository.GetUserRolesAsync(ownerId);
            if (!roles.Any(r => r.Name == "PrincipalInvestigator"))
                return null;

            var project = new Project
            {
                Title = request.Title,
                Description = request.Description,
                Objectives = request.Objectives,
                ExpectedOutcomes = request.ExpectedOutcomes,
                Status = "Draft",
                StartDate = request.StartDate,
                EndDate = request.EndDate,
                OwnerId = ownerId,
                ResearchTopicId = request.ResearchTopicId
            };

            var success = await _projectRepository.CreateAsync(project);
            if (!success)
                return null;

            // Add owner as a member with PI role
            await _projectRepository.AddMemberAsync(project.Id, ownerId, "PrincipalInvestigator");

            return new ProjectDto
            {
                Id = project.Id,
                Title = project.Title,
                Description = project.Description,
                Status = project.Status,
                StartDate = project.StartDate,
                EndDate = project.EndDate,
                OwnerId = project.OwnerId,
                OwnerName = owner.Name,
                ResearchTopicId = project.ResearchTopicId,
                CreatedAt = project.CreatedAt,
                UpdatedAt = project.UpdatedAt
            };
        }

        public async Task<bool> UpdateAsync(int id, UpdateProjectRequest request, int userId)
        {
            var project = await _projectRepository.GetByIdAsync(id);
            if (project == null)
                return false;

            // Check if user is the owner or has admin role
            if (project.OwnerId != userId)
            {
                var roles = await _userRepository.GetUserRolesAsync(userId);
                if (!roles.Any(r => r.Name == "Admin"))
                    return false;
            }

            project.Title = request.Title;
            project.Description = request.Description;
            project.Objectives = request.Objectives;
            project.ExpectedOutcomes = request.ExpectedOutcomes;
            project.Status = request.Status;
            project.StartDate = request.StartDate;
            project.EndDate = request.EndDate;
            project.ResearchTopicId = request.ResearchTopicId;

            return await _projectRepository.UpdateAsync(project);
        }

        public async Task<bool> DeleteAsync(int id, int userId)
        {
            var project = await _projectRepository.GetByIdAsync(id);
            if (project == null)
                return false;

            // Check if user is the owner or has admin role
            if (project.OwnerId != userId)
            {
                var roles = await _userRepository.GetUserRolesAsync(userId);
                if (!roles.Any(r => r.Name == "Admin"))
                    return false;
            }

            return await _projectRepository.DeleteAsync(id);
        }

        public async Task<bool> AddMemberAsync(int projectId, AddProjectMemberRequest request, int userId)
        {
            var project = await _projectRepository.GetByIdAsync(projectId);
            if (project == null)
                return false;

            // Check if user is the owner or has admin role
            if (project.OwnerId != userId)
            {
                var roles = await _userRepository.GetUserRolesAsync(userId);
                if (!roles.Any(r => r.Name == "Admin"))
                    return false;
            }

            // Check if member exists
            var member = await _userRepository.GetByIdAsync(request.UserId);
            if (member == null)
                return false;

            var success = await _projectRepository.AddMemberAsync(projectId, request.UserId, request.Role);
            if (success)
            {
                // Create notification for the new member
                var notification = new Notification
                {
                    Title = "Added to Project",
                    Message = $"You have been added to the project '{project.Title}' as {request.Role}.",
                    Type = "Info",
                    RelatedEntityType = "Project",
                    RelatedEntityId = projectId,
                    UserId = request.UserId
                };

                await _notificationRepository.CreateAsync(notification);
            }

            return success;
        }

        public async Task<bool> RemoveMemberAsync(int projectId, int memberId, int userId)
        {
            var project = await _projectRepository.GetByIdAsync(projectId);
            if (project == null)
                return false;

            // Check if user is the owner or has admin role
            if (project.OwnerId != userId)
            {
                var roles = await _userRepository.GetUserRolesAsync(userId);
                if (!roles.Any(r => r.Name == "Admin"))
                    return false;
            }

            // Cannot remove the owner
            if (project.OwnerId == memberId)
                return false;

            var success = await _projectRepository.RemoveMemberAsync(projectId, memberId);
            if (success)
            {
                // Create notification for the removed member
                var notification = new Notification
                {
                    Title = "Removed from Project",
                    Message = $"You have been removed from the project '{project.Title}'.",
                    Type = "Info",
                    RelatedEntityType = "Project",
                    RelatedEntityId = projectId,
                    UserId = memberId
                };

                await _notificationRepository.CreateAsync(notification);
            }

            return success;
        }

        public async Task<IEnumerable<UserDto>> GetProjectMembersAsync(int projectId)
        {
            var members = await _projectRepository.GetProjectMembersAsync(projectId);
            var memberDtos = new List<UserDto>();

            foreach (var member in members)
            {
                var roles = await _userRepository.GetUserRolesAsync(member.Id);

                memberDtos.Add(new UserDto
                {
                    Id = member.Id,
                    Email = member.Email,
                    Name = member.Name,
                    AvatarUrl = member.AvatarUrl,
                    Roles = roles.Select(r => r.Name).ToList()
                });
            }

            return memberDtos;
        }
    }
}

File: src/services/projectService.ts
import api from './apiService';
import { AxiosResponse } from 'axios';

export interface Project {
  id: number;
  title: string;
  proposerName?: string;
  description: string;
  objectives: string;
  expectedOutcomes: string;
  startDate: string;
  endDate: string;
  researchTopicId: number;
  ownerId?: number;
  status?: string;
  createdAt?: string;
  updatedAt?: string;
}

interface CreateProjectRequest {
  title: string;
  description: string;
  objectives: string;
  expectedOutcomes: string;
  startDate: string;
  endDate: string;
  researchTopicId: number;
}

interface UpdateProjectRequest extends Partial<CreateProjectRequest> {}

interface ProjectMember {
  id: number;
  userId: number;
  projectId: number;
  role?: string;
  joinedAt?: string;
}

interface AddMemberRequest {
  userId: number;
  role?: string;
}

// Error handler function
const handleError = (error: any, operation: string) => {
  console.error(`Error in ${operation}:`, error);
  
  if (error.response) {
    const { status, data } = error.response;
    throw new Error(`${operation} failed: ${data?.message || `HTTP ${status}`}`);
  } else if (error.request) {
    throw new Error(`${operation} failed: No response from server`);
  } else {
    throw new Error(`${operation} failed: ${error.message}`);
  }
};

// Lấy Pj bằng id
export const getProject = async (id: number): Promise<Project> => {
  try {
    const response: AxiosResponse<Project> = await api.get(`/Projects/${id}`);
    return response.data;
  } catch (error) {
    handleError(error, 'Get project');
    throw error;
  }
};

// Cập nhật Pj
export const updateProject = async (id: number, data: UpdateProjectRequest): Promise<Project> => {
  try {
    const response: AxiosResponse<Project> = await api.put(`/Projects/${id}`, data);
    return response.data;
  } catch (error) {
    handleError(error, 'Update project');
    throw error;
  }
};

// Xóa Pj
export const deleteProject = async (id: number): Promise<void> => {
  try {
    await api.delete(`/Projects/${id}`);
  } catch (error) {
    handleError(error, 'Delete project');
    throw error;
  }
};

// Lấy all Pj
export const getProjects = async (): Promise<Project[]> => {
  try {
    const response: AxiosResponse<Project[]> = await api.get('/Projects');
    return response.data;
  } catch (error) {
    handleError(error, 'Get projects');
    throw error;
  }
};

// Tạo Pj
export const createProject = async (data: CreateProjectRequest): Promise<Project> => {
  try {
    const response: AxiosResponse<Project> = await api.post('/Projects', data);
    return response.data;
  } catch (error) {
    handleError(error, 'Create project');
    throw error;
  }
};

//  dự án theo ID chủ sở hữu
export const getProjectsByOwner = async (ownerId: number): Promise<Project[]> => {
  try {
    const response: AxiosResponse<Project[]> = await api.get(`/Projects/owner/${ownerId}`);
    return response.data;
  } catch (error) {
    handleError(error, 'Get projects by owner');
    throw error;
  }
};

// Lấy project theo ID thành viên
export const getProjectsByMember = async (memberId: number): Promise<Project[]> => {
  try {
    const response: AxiosResponse<Project[]> = await api.get(`/Projects/member/${memberId}`);
    return response.data;
  } catch (error) {
    handleError(error, 'Get projects by member');
    throw error;
  }
};

// lấy trạng thái Pj
export const getProjectsByStatus = async (status: string): Promise<Project[]> => {
  try {
    const response: AxiosResponse<Project[] | { data: Project[] }> = await api.get(`/Projects/status/${status}`);
    if (Array.isArray(response.data)) return response.data;
    if (response.data && Array.isArray((response.data as any).data)) return (response.data as any).data;
    return [];
  } catch (error) {
    handleError(error, 'Get projects by status');
    throw error;
  }
};

// Lấy dự án theo research topic ID
export const getProjectsByResearchTopic = async (researchTopicId: number): Promise<Project[]> => {
  try {
    const response: AxiosResponse<Project[]> = await api.get(`/Projects/research-topic/${researchTopicId}`);
    return response.data;
  } catch (error) {
    handleError(error, 'Get projects by research topic');
    throw error;
  }
};

// Thêm thành viên vào Pj
export const addMemberToProject = async (projectId: number, memberData: AddMemberRequest): Promise<ProjectMember> => {
  try {
    const response: AxiosResponse<ProjectMember> = await api.post(`/Projects/${projectId}/members`, memberData);
    return response.data;
  } catch (error) {
    handleError(error, 'Add member to project');
    throw error;
  }
};

// Lấy thành viên dự án
export const getProjectMembers = async (projectId: number): Promise<ProjectMember[]> => {
  try {
    const response: AxiosResponse<ProjectMember[]> = await api.get(`/Projects/${projectId}/members`);
    return response.data;
  } catch (error) {
    handleError(error, 'Get project members');
    throw error;
  }
};

// Xóa thành viên ra khỏi Pj
export const removeMemberFromProject = async (projectId: number, memberId: number): Promise<void> => {
  try {
    await api.delete(`/api/Projects/${projectId}/members/${memberId}`);
  } catch (error) {
    handleError(error, 'Remove member from project');
    throw error;
  }
};

// PATCH /Projects/{id}/status - cập nhật trạng thái dự án
export const updateProjectStatus = async (id: number, status: string): Promise<void> => {
  try {
    await api.patch(`/api/Projects/${id}/status`, { status });
  } catch (error) {
    handleError(error, 'Update project status');
    throw error;
  }
};